vault/resource_pki_secret_backend_key.go

```diff
@@ -0,0 +1,265 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package vault
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"regexp"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/hashicorp/terraform-provider-vault/internal/consts"
+	"github.com/hashicorp/terraform-provider-vault/internal/provider"
+)
+
+var (
+	kmsType = "kms"
+
+	pkiSecretMountFromPathRegex = regexp.MustCompile("^(.+)/key/.+$")
+	pkiSecretKeyIDFromPathRegex = regexp.MustCompile("^.+/key/(.+)$")
+)
+
+func pkiSecretBackendKeyResource() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: MountCreateContextWrapper(pkiSecretBackendKeyCreate, provider.VaultVersion111),
+		UpdateContext: pkiSecretBackendKeyUpdate,
+		DeleteContext: pkiSecretBackendKeyDelete,
+		ReadContext:   ReadContextWrapper(ReadContextWrapper(pkiSecretBackendKeyRead)),
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			consts.FieldMount: {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Full path where PKI engine is mounted.",
+			},
+			consts.FieldType: {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Specifies the type of the key to create.",
+			},
+			consts.FieldKeyName: {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "When a new key is created with this request, optionally specifies the name for this.",
+			},
+			consts.FieldKeyType: {
```

Comment:
I notice you did not set a default here. Should we not be setting resource defaults in the schema in general?

Reply:
This is a great question! (and it also helped resolve a potential bug case, so thanks 🙇🏼) 

I'm sure there have been different styles for this as well so this is also a great spot for setting a pattern/precedent. My thoughts around this are that if there are defaults for fields that are enforced by Vault, then we don't need to set them in the TFVP and can instead also mark that field as being `Computed`. In this case, we can not write data to Vault for this field and let Vault set the default value. Since we marked it as `Computed` we will then obtain Vault's default value upon a read and then set it to the TF state. 

If Vault does not enforce the defaults itself, i.e., Vault does not set the defaults for fields we do not provide the data for, then the TFVP should set any defaults so that operations can still work. For example, in the case of this resource, Vault does not set the defaults for `key_type`. If the user does not provide a value for this 'optional' field, Vault's validation fails. In this case I think it would make sense for the TFVP to set defaults for schema fields. Ideally if the Vault docs mention a default then Vault should handle enforcing it, but Vault also seems inconsistent with thi behavior 😅 

Curious to hear what you think of the above! Thanks for pointing this out :)

Reply:
Oh, I assumed the default was `rsa` as per the [API docs](https://developer.hashicorp.com/vault/api-docs/secret/pki#key_type). So are the docs incorrect in this case?

Reply:
> if there are defaults for fields that are enforced by Vault, then we don't need to set them in the TFVP and can instead also mark that field as being Computed

Makes sense to me, I agree with that. Just to be clear, in this case you are recommending to not set the `Default` field?

> If Vault does not enforce the defaults itself, i.e., Vault does not set the defaults for fields we do not provide the data for, then the TFVP should set any defaults so that operations can still work.

I am not sure I agree with this part but it may be due to inexperience with TFVP here. My intuition is that we should not be setting a default value for Vault. If Vault doesn't return a default value and the user does not set a field that is required, why not let the operation fail?